search/includes/classes/commands/class-corecommand.php

```diff
@@ -14,38 +14,105 @@ class CoreCommand extends \ElasticPress\Command {
+			return WP_CLI::error( 'The --indexables argument is required when specifying --version, as each indexable has separate versioning' );
+		}
+
+		if ( array_key_exists( 'using-versions', $assoc_args ) && ! isset( $assoc_args['indexables'] ) ) {
+			return WP_CLI::error( 'The --indexables argument is required when specifying --using-versions' );
+		}
+
+		if ( array_key_exists( 'using-versions', $assoc_args ) && isset( $assoc_args['network-wide'] ) ) {
+			return WP_CLI::error( 'The --using-versions argument is not supported together with --network-wide' );
+		}
+	}
+
+	private function _shift_version_after_index( $assoc_args ) {
+		$search = \Automattic\VIP\Search\Search::instance();
+
+		$indexables = $this->_parse_indexables( $assoc_args );
+
+		foreach ( $indexables as $indexable ) {
+			WP_CLI::line( sprintf( 'Updating active version for "%s"', $indexable->slug ) );
+			$result = $search->versioning->activate_version( $indexable, 'next' );
+			if ( is_wp_error( $result ) ) {
+				return WP_CLI::error( sprintf( 'Error activating next version: %s', $result->get_error_message() ) );
+			}
+
+			WP_CLI::line( sprintf( 'Removing inactive version for "%s"', $indexable->slug ) );
+			$result = $search->versioning->delete_version( $indexable, 'previous' );
+			if ( is_wp_error( $result ) ) {
+				return WP_CLI::error( sprintf( 'Error deleting previous version: %s', $result->get_error_message() ) );
+			}
+		}
+	}
+
+	private function _parse_indexables( $assoc_args ) {
+		$indexable_slugs = explode( ',', str_replace( ' ', '', $assoc_args['indexables'] ) );
+
+		$indexables = [];
 
+		foreach ( $indexable_slugs as $slug ) {
+			$indexable = \ElasticPress\Indexables::factory()->get( $slug );
+
+			if ( ! $indexable ) {
+				return WP_CLI::error( sprintf( 'Indexable %s not found - is the feature active?', $slug ) );
+			}
+
+			$indexables[] = $indexable;
+		}
+		return $indexables;
+	}
+
+	protected function _maybe_setup_index_version( $assoc_args ) {
+		if ( array_key_exists( 'version', $assoc_args ) || array_key_exists( 'using-versions', $assoc_args ) ) {
+			$version_number = '';
+			$using_versions = $assoc_args['using-versions'] ?? false;
+			if ( $assoc_args['version'] ?? false ) {
+				$version_number = $assoc_args['version'];
+			} else if ( $using_versions ) {
+				$version_number = 'next';
+			}
+
+			if ( $version_number ) {
 				$search = \Automattic\VIP\Search\Search::instance();
 
 				// For each indexable specified, override the version
-				$indexable_slugs = explode( ',', str_replace( ' ', '', $assoc_args['indexables'] ) );
-
-				foreach ( $indexable_slugs as $slug ) {
-					$indexable = \ElasticPress\Indexables::factory()->get( $slug );
+				$indexables = $this->_parse_indexables( $assoc_args );
+
+				if ( $using_versions ) {
+					foreach ( $indexables as $indexable ) {
+						$current_versions = $search->versioning->get_versions( $indexable );
+						if ( count( $current_versions ) > 1 ) {
```

Comment:
I think I get the intent here, but I'm not sure whether we should be restrictive here. I.e. if there were multiple index versions I would still appreciate the ability to run this logic.  As a user, I'd like the ability to safely re-index regardless of how many old versions I have and I might be cautious enough as to not try messing with other index versions because I may not have enough context to know why they're there.

Reply:
It would actually fail without this check.

Firstly because we only allow 2 versions per indexable - https://github.com/Automattic/vip-go-mu-plugins/blob/b515aae5fe64d42f3bd924b37b466dbc20f3efb5/search/includes/classes/class-versioning.php#L21

And even if we would get around this. The current code is using "next" and "previous" to refer to the indexes when choosing which version to index, remove and activate. Which would break, if the user would already have "next" index before we start the process. I guess we could use numbers instead but that makes it slightly more complex. 

(mostly the first reason though)